Content Security Policy (CSP) (#1168)
snowpackjs/rfcs: proposals/0055-csp.md

<!--
  Note: You are probably looking for `stage-1--discussion-template.md`!
  This template is reserved for anyone championing an already-approved proposal.

  Community members who would like to propose an idea or feature should begin
  by creating a GitHub Discussion. See the repo README.md for more info.

  To use this template: create a new, empty file in the repo under `proposals/${ID}.md`.
  Replace `${ID}` with the official accepted proposal ID, found in the GitHub Issue
  of the accepted proposal.
-->

**If you have feedback and the feature is released as experimental, please leave it on the Stage 3 PR. Otherwise, comment on the Stage 2 issue (links below).**

- Start Date: 2025-05-13
- Reference Issues: 
- Implementation PR: 
- Stage 2 Issue: https://github.com/withastro/roadmap/issues/1149
- Stage 3 PR: https://github.com/withastro/roadmap/discussions/377

# Summary

> Add a new feature that allows Astro assets, mainly scripts and styles, to support [Content Security Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/CSP).

Reviewer: I don't think we'll handle anything else for now?
```suggestion
Add a new feature that allows Astro scripts and styles to support [Content Security Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/CSP).
```

# Example

```js
// astro.config.mjs
import { defineConfig } from "astro/config";

export default defineConfig({
    security: {
        csp: true
    }
})
```

# Background & Motivation

Include any useful background detail that that explains why this RFC is important.
What are the problems that this RFC sets out to solve? Why now? Be brief!

It can be useful to illustrate your RFC as a user problem in this section.

> (ex: "Users have reported that it is difficult to do X in Astro today.")

Reviewer: This is from the default template, we can probably extract things from the original discussion

# Goals

- Support at least one solution (`nonce`, hash, etc.) that would support CSP.
- Allow Astro scripts (client islands, server islands, view transitions, etc.) and Astro styles to work out of the box.
- Update adapters to support the feature, if needed.
- Make the feature opt-in, even when it's outside experimental.
- Support the [dev server](#dev-server) (stretch)
- View transitions (limited/opt-in support)

# Non-Goals

- Provide *multiple* solutions to fix the same problem
- Support for third-party scripts dynamically imported by users e.g. `document.createElement("script")`, `import("https://path.to/script.js")`;

> - Support for third-party styles dynamically imported by users.

Reviewer: If I undestand correctly, `injectCspAsset` would allow this

Author: I want to rephrase it then. What I meant is that **out of the box** Astro won't create a hash for the dynamically imported script/style. Any suggestions?

Reviewer: ```suggestion
- Automatic support for third-party styles dynamically imported by users.
```

- Performance. Users and authors are willing to overlook possible performance drawbacks in order to increase security.

# Detailed Design

## Solution adopted

Contrary to other frameworks of the same ecosystem, our solution will rely on generating [**hashes of styles and scripts**](https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/CSP#hashes). For more information, please refer to the [alternatives](#alternatives) section.

Astro provides the hashes during the rendering phase, they will be written in the `<head>` tag using the `<meta http-equiv="Content-Security-Policy">` meta tag.

The *main* reason for using this strategy instead of `Response` headers is because this solution allows Astro to support two important use cases:
- SSG, especially when no adapters are involved.
- SPA (Single Page Applications), where scripts and styles *might* be loaded dynamically using client-side routing.

## Generation of hashes

The hashes will be generated using the [`crypto`](https://developer.mozilla.org/en-US/docs/Web/API/Crypto) global object, and the package [`@oslojs/encoding`](https://encoding.oslojs.dev/). Astro *already uses* this package to encrypt props for server islands.

> Hashes will be generated using the `SHA-256` algorithm. Users can't change the algorithm.

Reviewer 2: I would advise against hard coding SHA-256 as SHA-384 or SHA-512 are already deployed in the wild.
Following CSP spec would be a better option

https://developer.mozilla.org/en-US/docs/Glossary/Hash_function

Author: Are you suggesting that users provide their own function? Or are you suggesting something else?

Reviewer 2: > Are you suggesting that users provide their own function? Or are you suggesting something else?

I'm suggesting that CSP integration should provide SHA-256, SHA-384 and SHA-512 as options for algorithm as they are pretty much standard across web dev space.

Also, I see now that `styleHashes` and `scriptHashes` sha256 requirement is being dropped, so I'm happy about it 👍

            algorithm: "SHA-512",
            // Applied only to `script-src` directive
            strictDynamic: true,
            // Additional directives added to the final content
            directives: [
                "default-src: 'self'",
                "img-src 'self'"
            ],
            // List of hashes for `style-src`. 
            styleHashes: [
                "sha512-hash1",
                "sha256-hash1",
                "sha384-hash1",
            ],
            // List of hashes for `script-src`. 
            scriptHashes: [
                "sha512-hash1",
                "sha256-hash1",
                "sha384-hash1",
            ],
        }
    }
})
```

### Runtime APIs

The following paragraph will describe possible runtime APIs that users can use to customise their CSP policies. When `Astro.` is used, it's intended that also `APIContext` is covered by the implementation.

#### `Astro.insertCspDirective(directive: string)`

Users can add an entire directive. Astro won't check if the directive is grammatically correct.

```js
Astro.insertCspPolicy(`image-src '${Astro.site}'`)
```
Result
```html
<meta 
  http-equiv="Content-Security-Policy"

>   content="image-src 'https://example.com'; style-src 'self' 'sha256-hash1'; script-src 'self' 'sha256-hash2'"

Reviewer: Does order matter?

Author: Can you be more specific with your query? The order of what :)

Reviewer: Sorry! I meant does it matter if it's first image-src, then style-src, then script-src or can it be in any order? eg. image-src at the end

Reviewer 2: It won't matter

>
```

#### `Astro.insertCspAsset({ content: string, kind: "script" | "style" })`

```js
Astro.insertCspAsset({
    content: "console.log(sort)",
    kind: "script"
})
```
Result
```html
<meta 
  http-equiv="Content-Security-Policy"
  content="style-src 'self' 'sha256-hash1'; script-src 'self' 'sha256-hash2' 'sha256-<NEW_HASH>'"
>
```
Where `<NEW_HASH>` is the hash generated by Astro from the `console.log(sort)` content. 

# Testing Strategy

Types of tests:
- unit tests
- integration tests
- e2e tests

Features to test:
- client islands
- server islands
- view transitions
- external scripts (via APIs)
- external styles (via APIs)
- inline scripts
- inline styles

# Drawbacks

> The hashes solution will negatively impact the rendering engine because Astro will have to calculate the hashes **at runtime**. The most expensive solution is the support of server islands because the contents of their scripts depend on props and slots, which are only known during the rendering phase.

Author: No, but we can set up a codspeed benchmark

> The hashes solution will negatively impact the rendering engine because Astro will have to calculate the hashes **at runtime**. The most expensive solution is the support of server islands because the contents of their scripts depend on props and slots, which are only known during the rendering phase.

Reviewer 2: On a comparable solution, it would take around 500ms for dozens of pages in a Netlify Free tier build. This should be the base performance

## Rendered `<meta>`

The current proposal targets scripts and styles emitted by Astro, so the directives rendered will handle only `style-src` and `script-src`:
```html
<meta 
  http-equiv="Content-Security-Policy"
  content="style-src 'self' 'sha256-hash1'; script-src 'self' 'sha256-hash2'"
>
```
While browsers are able to [merge directives](https://centralcsp.com/articles/csp-meta-tags) from `Response` headers and `meta` tags, Astro will provide runtime APIs so users can push their custom policies based in their requirements.

### Configuration APIs

New configurations will be available in case directives and/or hashes can be applied to all pages

```js
export default defineConfig({
    security: {
        csp: {
            // or SHA-384, or SHA-256 (default)
            algorithm: "SHA-512",
            // Additional directives added to the final content
            directives: [
                {
                    type: "default-src",

>                     content: "'sef'" 

Reviewer 2: Typo
```suggestion
                    content: "'self'" 
```

- Update adapters to support the feature, if needed.
- Make the feature opt-in, even when it's outside experimental.
- Support the [dev server](#dev-server) (stretch)
- View transitions (limited/opt-in support)

# Non-Goals

- Provide *multiple* solutions to fix the same problem
- Support for third-party scripts dynamically imported by users e.g. `document.createElement("script")`, `import("https://path.to/script.js")`;
- Support for third-party styles dynamically imported by users.
- Performance. Users and authors are willing to overlook possible performance drawbacks in order to increase security.

# Detailed Design

## Solution adopted

Contrary to other frameworks of the same ecosystem, our solution will rely on generating [**hashes of styles and scripts**](https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/CSP#hashes). For more information, please refer to the [alternatives](#alternatives) section.

Astro provides the hashes during the rendering phase, they will be written in the `<head>` tag using the `<meta http-equiv="Content-Security-Policy">` meta tag.

> The *main* reason for using this strategy instead of `Response` headers is because this solution allows Astro to support two important use cases:

Author: That's unsettling. Isn't it possible to use nonce and hashes at the same time?

> The *main* reason for using this strategy instead of `Response` headers is because this solution allows Astro to support two important use cases:

Reviewer 3: What I mean, you are truly doing a good job with the Astro out of the box solution. 

However what I am telling I do not want to use the meta tags implementation because it is an issue, like @Johannes-Andersen mentioned.

I already have my CSP strict dynamic setup through middleware. Using astro.locals.nonce setup. I inject the nonce now in  my scripts however than I cannot use typescript anymore very annoying but hey that's life :-) 

Then I really wanted to use preact for the island behaviour and than I noticed that my CSP messed up life. So I start trying to make a work-around which is not working yet. I use output = server so my site is dynamic and I really don't want to loose performance. Calculating hashes doesn't work because I have no static pages. I used to have a blog the hashes integration only calculated them for the static pages. So I tried middleware to adjust the generated html and insert nonces but that is not working as it should be.

It would be nice to setup a nonce which is injected by astro itself on page generation. Or maybe someone can help me to make the solution.

But I really appreciate your efforts.

> The *main* reason for using this strategy instead of `Response` headers is because this solution allows Astro to support two important use cases:

Reviewer 4: @NepCono our implementation is current based on hashes, why do you want to use nonce instead?

> The *main* reason for using this strategy instead of `Response` headers is because this solution allows Astro to support two important use cases:

Reviewer 3: Because the current hash implementation is through meta tags implementation, where somethings are not working.

Than I use a middleware implementation that I build myself to set to corresponding CSP response header.

With the current implementation of hashes being calculated everytime, it is far better and faster to use a nonce token. You only have to inject the nonce token in the scripts and your done (Astro.locals.nonce) when a user sets that problem solved. No hash calculation slowing things down, but hey that's only my opnion.

Also what @Johannes-Andersen is referring to.

In the mean time I am stuck, can you supply me with an astro integration that scans my pages (output server) so dynamic pages and or inject or hashes so I can move on ?

                    "sha512-hash1",
                    "sha256-hash1",
                    "sha384-hash1",
                ]
            },
            // customise the `style-src` directive
            scriptDirective: {
                // A list of resources that will override the default resources of Astro
                resources: [
                    "self",
                    "https://scripts.cdn.example.com",
                ],
                // List of hashes for `script-src`. 
                hashes: [
                    "sha512-hash1",
                    "sha256-hash1",
                    "sha384-hash1",
                    
                ]
            }
        }
    }
})
```

#### Provide additional directives

User can add additional directives to control other resources other than `script-src` and `style-src`.

These directives is a list of strings. Each string has this form

```ts
type CspDirective = `${AllowedDirectives} ${string}`;
```

Where `AllowedDirectives` is a list of the following directives:

```ts
export const ALLOWED_DIRECTIVES = [
	'base-uri',
	'child-src',
	'connect-src',
	'default-src',
	'fenced-frame-src',
	'font-src',
	'form-action',

> 	'frame-ancestors',

Author: With Astro v5.9.4, we shipped an adapter feature that allows to ship static headers using the hosting platform. As for now, only the Netlify adapter supports that: https://docs.astro.build/en/reference/adapter-reference/#experimentalstaticheaders

This should allow us to ship CSP headers via `Response` headers for static pages.

	'frame-src',
	'img-src',
	'manifest-src',
	'media-src',
	'object-src',
	'referrer',
	'report-to',
	'require-trusted-types-for',
	'sandbox',
	'trusted-types',
	'upgrade-insecure-requests',
	'worker-src',
] as const;
```

Here's an example where we add `default-src` directive amd `img-src` directive:

```js
export default defineConfig({
    security: {
        csp: {
            // Additional directives added to the final content
            directives: [
                "default-src 'self'",
                "img-src 'self' 'https://images.cdn.example.com'"
            ]
        }
    }
})
```

Will result in something similar to the following (the result is formatted to allow better readability):

```html
<meta 
  http-equiv="Content-Security-Policy"
  content="
    default-src 'self';

>     img-src 'self' 'https://images.cdn.example.com';

Reviewer 5: This would be invalid. Ref: https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Content-Security-Policy#host-source

```suggestion
    img-src 'self' https://images.cdn.example.com;
```

<img width="786" alt="Screenshot 2025-06-06 at 00 19 22" src="https://github.com/user-attachments/assets/799bcbc4-01b0-4515-b9fe-49afb28a2562" />

 "
>
```

### Runtime APIs

The following paragraph will describe possible runtime APIs that users can use to customise their CSP policies. When `Astro.` is used, it's intended that also `APIContext` is covered by the implementation.

#### `Astro.insertDirective(directive: CspDirective)`

Users can add an entire directive. Astro won't check if the directive is grammatically correct.

```js
Astro.insertDirective(`image-src '${Astro.site}'`)
```
Result
```html
<meta 
  http-equiv="Content-Security-Policy"
  content="image-src 'https://example.com'; style-src 'self' 'sha256-hash1'; script-src 'self' 'sha256-hash2'"
>
```

#### `Astro.insertStyleHash(string)`

```js
Astro.insertStyleHash("sha256-qewpiorfuvng3")
```
Result
```html
<meta 
  http-equiv="Content-Security-Policy"
  content="style-src 'self' 'sha256-hash1' 'sha256-qewpiorfuvng3'; script-src 'self' 'sha256-hash2'"
>
```

#### `Astro.setStyleResource(string)`

```js
Astro.setStyleResource("'https://styles.cdn.example.com'")
```
Result
```html
<meta 
  http-equiv="Content-Security-Policy"
  content="style-src 'https://styles.cdn.example.com' 'sha256-hash1' 'sha256-qewpiorfuvng3'; script-src 'self' 'sha256-hash2'"
>
```

> #### `Astro.insertScriptHash(string)`

Author: I still think this API isn't useful, as it essentially creates some unusual usage. Let's say you have some `script.js` somewhere. What would happen is the following:

- The user must import `script.js` *somewhere*, and provide its content *as a string* to the `generateHash`
- The generated hash must be saved or sent somewhere
- Then the user must provide this hash to Astro via configuration or runtime
- This very script must be added to the application 

Let's wait and see.

- Update adapters to support the feature, if needed.
- Make the feature opt-in, even when it's outside experimental.
- Support the [dev server](#dev-server) (stretch)
- View transitions (limited/opt-in support)

# Non-Goals

- Provide *multiple* solutions to fix the same problem
- Support for third-party scripts dynamically imported by users e.g. `document.createElement("script")`, `import("https://path.to/script.js")`;
- Support for third-party styles dynamically imported by users.
- Performance. Users and authors are willing to overlook possible performance drawbacks in order to increase security.

# Detailed Design

## Solution adopted

Contrary to other frameworks of the same ecosystem, our solution will rely on generating [**hashes of styles and scripts**](https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/CSP#hashes). For more information, please refer to the [alternatives](#alternatives) section.

Astro provides the hashes during the rendering phase. These hashes will served differently based on different cases:
- For static pages, they will be served in the `<head>` tag using the `<meta http-equiv="Content-Security-Policy">` meta tag.

> - For dynamic pages, they will be served in the `content-security-policyt` header of the `Response` type.

Reviewer 6: ```suggestion
- For dynamic pages, they will be served in the `content-security-policy` header of the `Response`.
```

> A [new adapter feature](https://docs.astro.build/en/reference/adapter-reference/#adapter-features) will be provided, called `hostedCsp`. For adapters that support this feature, Astro **won't** serve the hashes for *static pages*`.

Author: I don't mind renaming. 

> The we could use it for other stuff too.

What do you mean? Do you plan to make this feature generic?

> A [new adapter feature](https://docs.astro.build/en/reference/adapter-reference/#adapter-features) will be provided, called `hostedCsp`. For adapters that support this feature, Astro **won't** serve the hashes for *static pages*`.

Reviewer 6: The thing we're asking is whether the adapter supports setting headers for static pages, not anythign CSP-specific. I don't htink we have any specific plans, but if we keep the feature name generic, we can use it in future if we have anything else that uses headers.

> A [new adapter feature](https://docs.astro.build/en/reference/adapter-reference/#adapter-features) will be provided, called `hostedCsp`. For adapters that support this feature, Astro **won't** serve the hashes for *static pages*`.

Author: Addressed here [`eb28bb3` (#1168)](https://github.com/withastro/roadmap/pull/1168/commits/eb28bb3530ef4fb0ac03848b5db9f2833fbb20d8)

#### `Astro.insertDirective(directive: CspDirective)`

Users can add an entire directive. Astro won't check if the directive is grammatically correct.

```js
Astro.insertDirective(`image-src '${Astro.site}'`)
```
Result
```html
<meta 
  http-equiv="Content-Security-Policy"
  content="image-src 'https://example.com'; style-src 'self' 'sha256-hash1'; script-src 'self' 'sha256-hash2'"
>
```

#### `Astro.insertStyleHash(string)`

```js
Astro.insertStyleHash("sha256-qewpiorfuvng3")
```
Result
```html
<meta 
  http-equiv="Content-Security-Policy"
  content="style-src 'self' 'sha256-hash1' 'sha256-qewpiorfuvng3'; script-src 'self' 'sha256-hash2'"
>
```

#### `Astro.setStyleResource(string)`

```js
Astro.setStyleResource("'https://styles.cdn.example.com'")
```
Result
```html
<meta 
  http-equiv="Content-Security-Policy"
  content="style-src 'https://styles.cdn.example.com' 'sha256-hash1' 'sha256-qewpiorfuvng3'; script-src 'self' 'sha256-hash2'"
>
```

#### `Astro.insertScriptHash(string)`

```js

> Astro.insertScriptHash("sha256-qewpiorfuvng3")

Reviewer 5: ```suggestion
Astro.insertScriptHash("'sha256-qewpiorfuvng3'")
```

```
Result
```html
<meta 
  http-equiv="Content-Security-Policy"
  content="style-src 'self' 'sha256-hash1'; script-src 'self' 'sha256-hash2' 'sha256-qewpiorfuvng3'"
>
```

#### `Astro.setScriptResource(string)`

```js
Astro.setScriptResource("'https://scripts.cdn.example.com'")
```
Result
```html
<meta 
  http-equiv="Content-Security-Policy"

>   content="style-src 'self' 'sha256-hash1' 'sha256-qewpiorfuvng3'; script-src 'https://scripts.cdn.example.com' 'sha256-hash2'"

Reviewer 5: ```suggestion
  content="style-src 'self' 'sha256-hash1' 'sha256-qewpiorfuvng3'; script-src https://scripts.cdn.example.com 'sha256-hash2'"
```

<!--
  Note: You are probably looking for `stage-1--discussion-template.md`!
  This template is reserved for anyone championing an already-approved proposal.

  Community members who would like to propose an idea or feature should begin
  by creating a GitHub Discussion. See the repo README.md for more info.

  To use this template: create a new, empty file in the repo under `proposals/${ID}.md`.
  Replace `${ID}` with the official accepted proposal ID, found in the GitHub Issue
  of the accepted proposal.
-->

**If you have feedback and the feature is released as experimental, please leave it on the Stage 3 PR. Otherwise, comment on the Stage 2 issue (links below).**

- Start Date: 2025-05-13
- Reference Issues: 

> - Implementation PR: 

Reviewer: ```suggestion
- Implementation PR: https://github.com/withastro/astro/pull/13802
```

Due to this significant discrepancy, supporting the dev server will be very difficult, so we will focus on the build first.

# Alternatives

## The `nonce` solution

This solution is way **easier** to implement; however, it won't work for static pages (SSG).
As highlighted in this [comment of mine](https://github.com/withastro/roadmap/issues/1149#issuecomment-2824552131), in order to provide the `nonce` header, a user would need to use **an edge function** in a serverless environment or middleware for server environments. This means that environments that don't provide any virtual host control (e.g. GitHub pages, etc.), the solution wouldn't work *at all*.

This solution doesn't follow Astro core principles, where an Astro site should work *out of the box* with the majority of environments, especially SSG, which is a first-class citizen for us.

# Adoption strategy

- The usage of this feature will be under an experimental flag, and shipped in a minor:

  ```js
  export default defineConfig({
      experimental: {
          csp: true
      }
  })
  ```

- The new Astro feature will be shipped with prefix `_exprimental`: `_experimentalStaticHeaders` in a patch release.

> - The new field of the hook `astro:build:generated` will be shipped with prefix `_experimental`: `_experimentalRouteToHeaders` in a patch release.

Reviewer: When you make a PR against `next` to stabilize CSP, we gotta remember to update these names

Author: `_experimentalRouteToHeaders` was added for CSP, but it's been designed to be generic. Which means that the two features aren't directly connected, and they can become stable independently. If by the time of stabilisation, `_experimentalRouteToHeaders` is still named like this, we shouldn't change it.
